src/Data/String/Regex.js

```diff
@@ -67,12 +67,21 @@ exports.replace = function (r) {
   };
 };
 
-exports.replaceBy = function (r) {
-  return function (f) {
-    return function (s2) {
-      return s2.replace(r, function (match) {
-        return f(match)(Array.prototype.splice.call(arguments, 1, arguments.length - 3));
-      });
+exports._replaceBy = function (just) {
+  return function (nothing) {
+    return function (r) {
+      return function (f) {
+        return function (s) {
+          return s.replace(r, function (match) {
+            var groups = [];
+            var group, i = 1;
+            while (typeof (group = arguments[i++]) !== "number") {
+              groups.push(group == null ? nothing : just(group));
+            }
```

Comment:
The `typeof` check looks odd, but is necessary to handle both possible cases:

  - `match, group1, group2, ..., offset, string` (`r` contains no named capturing groups); and
  - `match, group1, group2, ..., offset, string, groups` (`r` contains named capturing groups).